src/test/java/co/rsk/bitcoinj/core/TransactionWitnessTest.java

```diff
@@ -186,11 +202,231 @@ public void equals_withTwoTransactionWitnessesWithOneDifferentPush_shouldBeFalse
+    }
+
+    @Test
+    public void getSigInsertionIndex_withFedKey3SignatureInTheWitness_shouldReturnIndexCorrectly() {
+        // arrange
+        BtcTransaction btcTx = getBtcTransactionWithBaseWitnessInInput();
+        Sha256Hash btcTxSigHash = btcTx.hashForWitnessSignature(FIRST_INPUT_INDEX, redeemScript, prevValue,
+            BtcTransaction.SigHash.ALL, false);
+
+        // sign with fedKey3
+        signInput(btcTx, fedKey3, FIRST_INPUT_INDEX, btcTxSigHash);
+        TransactionWitness witnessWithFedKey3Signature = btcTx.getWitness(FIRST_INPUT_INDEX);
+
+        // act
+        int sigIndexForFedKey1 = witnessWithFedKey3Signature.getSigInsertionIndex(btcTxSigHash, fedKey1);
+        int sigIndexForFedKey2 = witnessWithFedKey3Signature.getSigInsertionIndex(btcTxSigHash, fedKey2);
+
+        // assert
+        Assert.assertEquals(0, sigIndexForFedKey1);
+        Assert.assertEquals(0, sigIndexForFedKey2);
+    }
+
+    @Test
+    public void getSigInsertionIndex_withWitnessFilledWithSignatures_shouldReturnTheProperIndex() {
+        BtcTransaction btcTx = getBtcTransactionWithBaseWitnessInInput();
+        Sha256Hash btcTxSigHash = btcTx.hashForWitnessSignature(FIRST_INPUT_INDEX, redeemScript, prevValue,
+            BtcTransaction.SigHash.ALL, false);
+        TransactionWitness transactionWitness = btcTx.getWitness(FIRST_INPUT_INDEX);
+
+        int witnessPushCount = transactionWitness.getPushCount();
+        int sigsSuffixCount = p2shOutputScript.getSigsSuffixCount();
+
+        // the pushes that should have the signatures
+        // are between first one (empty byte for checkmultisig bug)
+        // and second to last one (op_notif + redeem script)
+        byte[] secondToLastPush = transactionWitness.getPush(witnessPushCount - sigsSuffixCount - 1);
+        byte[] emptyByte = new byte[]{};
+
+        int i = 0;
+        while(Arrays.equals(secondToLastPush, emptyByte)) {
+            BtcECKey key = FEDERATION_KEYS.get(i);
+            signInput(btcTx, key, FIRST_INPUT_INDEX, btcTxSigHash);
+            TransactionWitness witnessWithSignatures = btcTx.getWitness(FIRST_INPUT_INDEX);
+            secondToLastPush = witnessWithSignatures.getPush(witnessPushCount - sigsSuffixCount - 1);
+            i++;
+        }
+
+        TransactionWitness signedTransactionWitness = btcTx.getWitness(FIRST_INPUT_INDEX);
+        BtcECKey key = FEDERATION_KEYS.get(i);
+        int sigInsertionIndex = signedTransactionWitness.getSigInsertionIndex(btcTxSigHash, key);
+        assertEquals(i, sigInsertionIndex);
+    }
+
+    private static BtcTransaction getPreviousBtcTransaction() {
+        BtcTransaction btcTx = new BtcTransaction(MAINNET_PARAMS);
+        final Address userAddress = BtcECKey.fromPrivate(BigInteger.valueOf(901)).toAddress(MAINNET_PARAMS);
+        btcTx.addOutput(prevValue, userAddress);
+        return btcTx;
+    }
+
+    private static BtcTransaction getBtcTransactionWithBaseWitnessInInput() {
```

Comment:
Let's put 2 or 3 inputs in the transaction just in case, to make it a bit more complete

Reply:
done